fedbiomed/researcher/requests/_requests.py

```diff
@@ -0,0 +1,495 @@
+REQUEST_STATUS_CHECK_TIMEOUT = 0.5
+
+
+class MessagesByNode(dict):
+    """Type to defined messages by node"""
+    pass
+
+
+class Request:
+
+    def __init__(
+        self,
+        message: Message,
+        node: NodeAgent,
+        request_id: Optional[str] = None,
+        sem_pending: threading.Semaphore = None
+    ) -> None:
+        """Single request for node
+
+        Args:
+            message: Message to send to the node
+            node: Node agent
+            request_id: unique ID of request
+            sem_pending: semaphore for signaling new pending reply
+        """
+        self.request_id = request_id if request_id else str(uuid.uuid4())
+        self.node = node
+        self.message = message
+
+        self._sem_pending = sem_pending
+
+        self.reply = None
+        self.error = None
+        self.status = None
+
+    def has_finished(self) -> bool:
+        """Queries if the request has finished.
+
+        Also tracks if node has disconnected.
+
+        Args:
+            True if a reply was received from node
+        """
+
+        if self.node.status == NodeActiveStatus.DISCONNECTED:
+            self.status = RequestStatus.DISCONNECT
+            return True  # Don't expect any reply
+
+        return True if self.reply or self.error else False
+
+    def send(self) -> None:
+        """Sends the request"""
+        self.message.request_id = self.request_id
+        self.node.send(self.message, self.on_reply)
+        self.status = RequestStatus.NO_REPLY_YET
+
+    def flush(self, stopped: bool) -> None:
+        """Flushes the reply that has been processed
+
+        Args:
+            stopped: True if the request was stopped before completion
+        """
+        self.node.flush(self.request_id, stopped)
+
+    def on_reply(self, reply: dict) -> None:
+        """Callback for node agent to execute once it replies.
+
+        Args:
+            reply: reply message received from node
+        """
+
+        if isinstance(reply, ErrorMessage):
+            self.error = reply
+            self.status = RequestStatus.ERROR
+        else:
+            self.reply = reply
+            self.status = RequestStatus.SUCCESS
+
+        self._sem_pending.release()
+
+
+class FederatedRequest:
+    """Dispatches federated requests
+
+    This class has been design to be send a request and wait until a
+    response is received
+    """
+    def __init__(
+        self,
+        message: Union[Message, MessagesByNode],
+        nodes: List[NodeAgent],
+        policy: Optional[List[RequestPolicy]] = None
+    ):
+        """Constructor of the class.
+
+        Args:
+            message: either a common `Message` to send to nodes or a dict of distinct message per node
+                indexed by the node ID
+            nodes: list of nodes that are sent the message
+            policy: list of policies for controlling the handling of the request
+        """
+
+        self.message = message
```

Comment:
Hi @srcansiz 
All variables from `FederatedRequests` should probably be private withe getter when needed (eg `self.policies` `self.requests`.

This would be nice implementation IMO as it prevents direct modification of the variables (which is not supposed to occur)